tidy3d/plugins/design/design.py

```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        console = get_logging_console() if verbose else None
 
-        fn_source = self.get_fn_source(function)
+        # Run based on how many functions the user provides
+        if fn_post is None:
+            fn_args, fn_values, aux_values, opt_output = self.run_single(fn, console)
+            sim_names = None
 
-        # package the result
-        return self._package_run_results(fn_args=fn_args, fn_values=fn_values, fn_source=fn_source)
+        else:
+            fn_args, fn_values, aux_values, opt_output, sim_names = self.run_pre_post(
+                fn_pre=fn, fn_post=fn_post, console=console
+            )
 
-    @staticmethod
-    def _make_batch_fn_source(fn_source_pre: str, fn_source_post: str) -> str:
-        """How to make the full function source from the pre and post functions."""
+            sim_names = tuple(sim_names)
 
-        if (fn_source_pre is None) and (fn_source_post is None):
-            return None
+            if len(sim_names) == 0:
+                sim_names = None
 
-        return str(fn_source_pre) + "\n\n" + str(fn_source_post)
+        fn_source = self.get_fn_source(fn)
+
+        # Package the result
+        return self._package_run_results(
+            fn_args=fn_args,
+            fn_values=fn_values,
+            fn_source=fn_source,
+            aux_values=aux_values,
+            task_ids=sim_names,
+            opt_output=opt_output,
+        )
+
+    def run_single(self, fn: Callable, console) -> Tuple(list[dict], list, list[Any]):
+        """Run a single function of parameter inputs."""
+        evaluate_fn = self._get_evaluate_fn_single(fn=fn)
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters, console=console)
+
+    def run_pre_post(self, fn_pre: Callable, fn_post: Callable, console) -> Tuple(
+        list[dict], list[dict], list[Any]
+    ):
+        """Run a function with tidy3d implicitly called in between."""
+        handler = self._get_evaluate_fn_pre_post(
+            fn_pre=fn_pre, fn_post=fn_post, fn_mid=self._fn_mid
+        )
+        fn_args, fn_values, aux_values, opt_output = self.method.run(
+            run_fn=handler.evaluate, parameters=self.parameters, console=console
+        )
+        return fn_args, fn_values, aux_values, opt_output, handler.sim_names
+
+    """ Helpers """
+
+    def _get_evaluate_fn_single(self, fn: Callable) -> list[Any]:
+        """Get function that sequentially evaluates single `fn` for a list of arguments."""
+
+        def evaluate(args_list: list) -> list[Any]:
+            """Evaluate a list of arguments passed to ``fn``."""
+            return [fn(**args) for args in args_list]
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable, fn_mid) -> list[Any]:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        class Pre_Post_Handler:
```

Comment:
> what's the difference between this vs. just writing `fn_combined` as a method and having the `counter` and `sim_names` as variables in that method?

Names have to be made within 'fn_mid', which sits within 'fn_combined', as they are used in the web.run calls to give sensible names to the servers. 

'fn_combined' can't store any variables because it has to be capable of being called once for the samplers, and being called many times by the optimizers. The optimizers call 'fn_combined' within their fitness functions which are very difficult to get anything into or out of. We need something that sits in the same scope as 'fn_combined' to capture names because we want to have these names for the Results object (although currently we're not doing much with them in Results, but it seems sensible to keep everything together). I'd rather not handle this for each method as it will bloat the method code and make adding new methods harder.

As to what the names are it's very open to interpretation. Currently the handler is also storing a counter and that is built into the name, but there are many different ways we could name things. What's the default naming conventions? Can the user not supply a task_id?

Reply:
> Names have to be made within fn_mid, which sits within fn_combined, as they are used in the web.run calls to give sensible names to the servers.

> fn_combined can't store any variables because it has to be capable of being called once for the samplers, and being called many times by the optimizers. The optimizers call fn_combined within their fitness functions which are very difficult to get anything into or out of. We need something that sits in the same scope as fn_combined to capture names because we want to have these names for the Results object (although currently we're not doing much with them in Results, but it seems sensible to keep everything together). I'd rather not handle this for each method as it will bloat the method code and make adding new methods harder.

Could the suffixes for example be passed from the method as it calls `fn_combined`? of course, they would need to be passed through to `fn_mid` and stuff. Would this cause more complication?

I think the current naming convention is ok, but it does sort of lock us into a specific one whereas allowing each method to pass whatever suffix (or even a function generating the task_name) might be more flexible? I dont understand all of the logic though, so I defer to your judgement on it.

> Can the user not supply a task_id?

Not sure what you mean? like the `Job.task_id`?